src/sinks/cratedb.rs

```diff
@@ -0,0 +1,546 @@
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct BasicAuth {
+    user: String,
+    password: String,
+}
+
+impl BasicAuth {
+    fn apply(&self, header_map: &mut http::header::HeaderMap) {
+        let auth = headers::Authorization::basic(&self.user, &self.password);
+        header_map.typed_insert(auth)
+    }
+}
+
+#[typetag::serde(name = "cratedb")]
+impl SinkConfig for CrateDBSinkConfig {
+    fn build(&self, acker: Acker) -> Result<(super::RouterSink, super::Healthcheck), String> {
+        let sink = cratedb(self.clone(), acker)?;
+
+        let healtcheck = healthcheck(self.host.clone(), self.basic_auth.clone())?;
+        Ok((sink, healtcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+#[derive(Debug)]
+enum BufferValue {
+    Statement(String),
+    BulkArgs(Vec<Vec<MapValue>>),
+}
+
+impl Serialize for BufferValue {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match &self {
+            BufferValue::Statement(s) => s.serialize(serializer),
+            BufferValue::BulkArgs(ba) => ba.serialize(serializer),
+        }
+    }
+}
+
+pub struct CrateDBBuffer {
+    statement: String,
+    inner: Vec<Vec<MapValue>>,
+}
+
+impl CrateDBBuffer {
+    pub fn new(statement: String) -> Self {
+        Self {
+            statement: statement.clone(),
+            inner: Vec::new(),
+        }
+    }
+}
+
+impl Batch for CrateDBBuffer {
+    type Input = Vec<MapValue>;
+    type Output = Vec<u8>;
+
+    fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    fn push(&mut self, item: Self::Input) {
+        self.inner.push(item)
+    }
+
+    fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    fn fresh(&self) -> Self {
+        Self {
+            statement: self.statement.clone(),
+            inner: Vec::new(),
+        }
+    }
+
+    fn finish(self) -> Self::Output {
+        let num_events = self.inner.len();
+        let body = json!({
+            "stmt": BufferValue::Statement(self.statement),
+            "bulk_args": BufferValue::BulkArgs(self.inner),
+        });
+        info!("Sending {} events", num_events);
+        let out = serde_json::to_vec(&body).unwrap_or(Vec::new());
+        out
+    }
+
+    fn num_items(&self) -> usize {
+        self.inner.num_items()
+    }
+}
+
+fn cratedb(config: CrateDBSinkConfig, acker: Acker) -> Result<super::RouterSink, String> {
+    let host = build_uri(&config.host)?;
+    let schema = config.schema.unwrap_or(String::from("doc"));
+    let table = config.table.unwrap_or(String::from("logging"));
+    let columns = config
+        .columns
+        .unwrap_or(vec![String::from("timestamp"), String::from("message")]);
+    let keys = config
+        .keys
+        .unwrap_or(vec![String::from("timestamp"), String::from("message")]);
+    let _params = vec!["?"; columns.len()];
```

Comment:
unused binding?

Reply:
No, used a few lines below. I need _n_ `?` joined by `,`: `_params.join(", ")`. I'm happy to drop the leading `_`.

Reply:
Done